charms/istio-pilot/src/charm.py

```diff
@@ -441,16 +444,38 @@ def _check_leader(self):
             raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
 
     @property
-    def _cert_subject(self):
-        """Return the domain to be used in the CSR."""
+    def _ingress_gateway_host(self) -> Optional[str]:
+        """Return a domain name if it's configured, the ingress gateway svc address otherwise."""
+        # If the configuration option is set, we use it
+        if domain_name := self.model.config["domain-name"]:
+            return domain_name
+
+        # We try getting the ingress gateway IP address in the absence of domain name
         try:
             svc = self._get_gateway_service()
         except ApiError:
-            self.log.info("Could not retrieve the gateway service for configuring the CSR.")
+            self.log.info("Could not retrieve the gateway service for using in the CSR.")
             return None
-        gateway_address = _get_gateway_address_from_svc(svc)
-        if gateway_address:
-            return gateway_address
+        return _get_gateway_address_from_svc(svc)
+
+    @property
+    def _cert_subject(self) -> Optional[str]:
+        """Return the certificate subject to be used in the CSR."""
+        target = self._ingress_gateway_host
+
+        # Check if the target is a hostname and return if it is
+        if _is_hostname(target):
+            assert isinstance(target, str), target
+            return target
+
+        # If target is an IP address, try to get the hostname and return it
+        with contextlib.suppress(OSError, TypeError):
+            name, _, _ = socket.gethostbyaddr(target)
```

Comment:
if I understand correctly, we get here when:
* domain-name is not set
* target (the gateway service's address) is an IP address

and then we try to lookup a hostname associated with that IP address.  Is that right?

If yes:
1. does this always work?  If it does, do we even need to let users specify domain-name manually, or can they rely on this alone?
2. if this only works sometimes, is it worth implementing at all?  Or should we just say "if you don't specify a domain name and we can't see it from the load_balancer, then you can't get a cert"?  

If (1) is not true, I suggest considering (2) to simplify the logic.  The "look up a host based on the IP address" behaviour feels a lot harder to predict edge cases for, and it also makes the code more complex to read.  If it really does add value then maybe its worth it, but I'm not sure whether it does.

Reply:
The reason why we need this is that often times the ingress gateway `Service` will point to an IP address, it rarely has an actual hostname, this can happen if users set their DNS server to make a domain name resolve to the Service IP address.

In a next iteration of this PR I will remove the configuration option and instead let the charm discover and construct this domain name using all this logic. Let me know what you think and if it is clearer after the update.